ReadMe.md
# youtube->bilibili 搬运脚本
![STARS](https://img.shields.io/github/stars/ame-yu/u2bili?color=%231cd&style=for-the-badge)
![LICENSE](https://img.shields.io/github/license/ame-yu/u2bili?style=for-the-badge)
![LAST_COMMIT](https://img.shields.io/github/last-commit/ame-yu/u2bili?style=for-the-badge)
<details>
    <summary>Demo预览</summary>

![Preview](docs/preview.gif)
</details>

## 🍔使用
> 预置环境 node+python+[jq](https://github.com/stedolan/jq)
> 
1. `yarn`或者`npm install`安装node依赖
2. 安装yt-dlp [文档](https://github.com/yt-dlp/yt-dlp#installation)
3. 提供cookies：编辑config.js或设置`BILIBILI_COOKIE`环境变量
4. bash u2bili.sh \<url\>

<details>
    <summary>关于获取Cookie</summary>
    
登录后F12,Application(应用程序)面板，选择cookie进行查看。
![Cookie](docs/cookie.jpg)
填写[config.js](config.js)最后4个参数或设置`BILIBILI_COOKIE`环境变量
```
BILIBILI_COOKIE环境变量格式如下：
DedeUserID=XXX; DedeUserID__ckMd5=XXX; bili_jct=XXX; SESSDATA=XXX
```
</details>

<details>
    <summary>✨字幕上传脚本预览</summary>
    
![Cookie](docs/preview_uploadsubs.gif)
</details>

## 🍱使用框架 Frameworks
- yt-dlp
- playwright
- actionsflow

## 🧂Q&A
<details>
    <summary>📺关于下载的清晰度</summary>

如有装有ffmpeg则会自动选择高画质视频和高画质音频然后合并。
yt-dlp文档 [github.com/yt-dlp/yt-dlp#format-selection](https://github.com/yt-dlp/yt-dlp#format-selection)
</details>

<details>
    <summary>🍥使用Github Action</summary>

<h2 ><b>❗重要提示：请clone后push到自己的私有仓库，使用<a href="https://github.com/settings/billing">额度内action时间！<a></b></h2>
<br>

Actions面板设置Secret `BILIBILI_COOKIE` （必要步骤）
```
DedeUserID=XXX; DedeUserID__ckMd5=XXX; bili_jct=XXX; SESSDATA=XXX
```
几个重要参数
- 扫描周期`schedule.cron` [.github/workflows/actionsflow.yml](.github/workflows/actionsflow.yml)
- 订阅频道`channel_id` [workflows/youtube.yml](workflows/youtube.yml)
- 视频条目过滤`filterScript` 默认只对比了时间选取24小时内的视频 [workflows/youtube.yml](workflows/youtube.yml)
  - [脚本文档](https://actionsflow.github.io/docs/workflow/#ontriggerconfigfilterscript)
  - [完整视频参数](https://actionsflow.github.io/docs/triggers/youtube/#outputs)
</details>

## ⚠免责声明
项目仅用于学习参考，如存在违反B站用户协议请使用者风险自负。

## 📜Licence
MIT



CHANGELOG
## 1.2.0 20220512
1. youtube-dl 换成 yt-dlp

## 1.1.0 20211023
1. 填坑，定期扫描视频自动上传字幕脚本
2. 补充上传字幕脚本使用文档

## 1.0.0 20210806 
1. 固定act运行的镜像，原catthehacker/ubuntu:act-20.04镜像node被升成了v14.17.4，和playwright存在兼容问题。导致元素找不到。现使用catthehacker/ubuntu:act-20.04-20210721
2. 不再使用原ActionsFlow的方法检查同时运行中的工作流，详情Issue：https://github.com/actionsflow/actionsflow/issues/28


init.sh
#!/usr/bin/env bash

# Template from https://betterdev.blog/minimal-safe-bash-script-template/

# Check enviroment

set -Eeuo pipefail
trap cleanup SIGINT SIGTERM ERR EXIT

script_dir=$(cd "$(dirname "${BASH_SOURCE[0]}")" &>/dev/null && pwd -P)

usage() {
  cat <<EOF
Usage: $(basename "${BASH_SOURCE[0]}") [-h] [-v] [-f] -p param_value arg1 [arg2...]

Script description here.

Available options:

-h, --help      Print this help and exit
-v, --verbose   Print script debug info
-f, --flag      Some flag description
-p, --param     Some param description
EOF
  exit
}

cleanup() {
  trap - SIGINT SIGTERM ERR EXIT
  # script cleanup here
}

setup_colors() {
  if [[ -t 2 ]] && [[ -z "${NO_COLOR-}" ]] && [[ "${TERM-}" != "dumb" ]]; then
    NOFORMAT='\033[0m' RED='\033[0;31m' GREEN='\033[0;32m' ORANGE='\033[0;33m' BLUE='\033[0;34m' PURPLE='\033[0;35m' CYAN='\033[0;36m' YELLOW='\033[1;33m'
  else
    NOFORMAT='' RED='' GREEN='' ORANGE='' BLUE='' PURPLE='' CYAN='' YELLOW=''
  fi
}

msg() {
  echo >&2 -e "${1-}"
}

die() {
  local msg=$1
  local code=${2-1} # default exit status 1
  msg "$msg"
  exit "$code"
}

parse_params() {
  # default values of variables set from params
  flag=0
  param=''

  while :; do
    case "${1-}" in
    -h | --help) usage ;;
    -v | --verbose) set -x ;;
    --no-color) NO_COLOR=1 ;;
    -f | --flag) flag=1 ;; # example flag
    -p | --param) # example named parameter
      param="${2-}"
      shift
      ;;
    -?*) die "Unknown option: $1" ;;
    *) break ;;
    esac
    shift
  done

  args=("$@")

  # check required params and arguments
  #[[ -z "${param-}" ]] && die "Missing required parameter: param"
  #[[ ${#args[@]} -eq 0 ]] && die "Missing script arguments"
  
  return 0
}

parse_params "$@"
setup_colors

# script logic here
# msg "Change to user: ${RED}$username${NOFORMAT}"
if ! (command -v node && command -v npm) &> /dev/null; then
    msg "${RED}× Node or npm could not be found"
    exit
fi

npm list | grep playwright || npm install playwright --no-shrinkwrap
msg "${GREEN}√ playwright installed"


if ! command -v pip &> /dev/null; then
    msg "${RED}× pip could not be found, "
fi

if ! command -v youtube-dl &> /dev/null; then
    msg "${YELLOW}! youtube-dl could not be found, try to install"
    pip install youtube-dl
fi
msg "${GREEN}√ youtube-dl installed"

if ! [[ -t downloads ]]; then
    mkdir downloads
fi

msg "${GREEN}√ Ready to go (2/2)"
sleep 1


upload.js
import {
  bilibiliCookies,
  showBrowser,
  downloadPath,
} from "./config.js"
import { firefox as browserCore } from "playwright"
import { existsSync, readFileSync } from "fs"
/**
 * 使用例 node upload.sh MetaFile [VideoFile]
 * MetaFile 是必须的
 * 如果视频文件命名和Meta文件一致则可不写
 */
const uploadPageUrl = "https://member.bilibili.com/video/upload.html"

if (process.argv.length < 3) {
  console.error("至少传入视频信息JSON路径 node upload.js json_file [video_file]")
  process.exit(-1)
}

var [metaPath, videoPath] = process.argv.slice(2, 4)
const meta = JSON.parse(readFileSync(metaPath))

function getCookies(){
  const envCookies = process.env["BILIBILI_COOKIE"]
  if(envCookies){
    console.log("从环境变量读取Cookie");
    return envCookies.split(";").map(i => {
      const [key, value] = i.split("=")
      return {
        domain: ".bilibili.com",
        path: "/",
        name: key,
        value: value,
      }
    })
  }else{
    console.log("从配置文件读取Cookie");
    return Object.keys(bilibiliCookies).map((k) => {
      return {
        domain: ".bilibili.com",
        path: "/",
        name: k,
        value: bilibiliCookies[k],
      }
    })
  }
}

async function main() {
  const browser = await browserCore.launch({
    headless: !showBrowser,
  })
  const context = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.182 Safari/537.36 Edg/88.0.705.74",
    storageState: {
      origins: [
        {
          origin: "https://member.bilibili.com",
          localStorage: [
            {
              name: "SHOW_GUIDE",
              value: "1",
            },
          ],
        },
      ],
    },
  })
  context.addCookies(getCookies())
  const page = await context.newPage()
  try {
    await Promise.all([
      page.goto(uploadPageUrl, { waitUntil: "networkidle", timeout: 20 * 1000 }),
      page.waitForResponse(/\/OK/), //Fix：库未加载完的无效点击
    ])
  } catch (error) {
    if (error.name === "TimeoutError") {
      console.log("超时! 当前页面:" + page.url())
    }
  }

  const [fileChooser] = await Promise.all([
    page.waitForEvent("filechooser"),
    page.click("#bili-upload-btn"),
  ])

  if (!videoPath) {
    const ext = ["webm", "mp4", "mkv"].find((ext) =>
      existsSync(`${downloadPath}${meta["id"]}.${ext}`)
    )
    if (!ext) {
      console.error(
        `无法在${downloadPath}找到${meta["id"]}命名的视频文件，上传未成功。`
      )
      process.exit(-1)
    }
    videoPath = `${downloadPath}${meta["id"]}.${ext}`
  }
  await fileChooser.setFiles(videoPath)
  console.log(`开始上传${videoPath}`);

  await page.click('text="转载"')
  await page.fill("input[placeholder^=转载视频请注明来源]", meta["webpage_url"])
  await page.fill("input[placeholder*=标题]", meta["title"])

  // 选择分区
  await page.click("div.select-box-v2-container")
  await page.click('text="知识"')
  await page.click("div.drop-cascader-list-wrp > div:nth-child(8)") // 修复问题:找不到二级选项导致堵塞，数字对应二级列表位置
  //await page.click('text="野生技术协会"')

  // 创建标签
  await page.click("input[placeholder*=创建标签]")
  await page.keyboard.type(meta["uploader"])
  await page.keyboard.down("Enter")

  // 视频描述
  await page.click("div.ql-editor[data-placeholder^=填写更全]")
  await page.keyboard.type(
    `u2bili自动上传\n${meta["description"]}`.slice(0, 250)
  )

  //更多选项
  await page.click('text="更多选项"')
  await page.click('text="允许观众投稿字幕"')

  await page.waitForSelector('text="上传完成"',{
    timeout: 5 * 60_000 //等待上传5分钟
  }).catch(() => {
    console.log("上传时间过长")
  })

  await page.click('text="立即投稿"')
  let result = await page.textContent("h3.upload-3-v2-success-hint-1",{
    timeout: 60_000
  })
  console.log(result)

  await page.waitForTimeout(3_000)
  await page.close()
  await context.close()
  await browser.close()
}

main()


uploadsubs.js
import fs from "fs"
import https from "https"
import { bilibiliCookies, downloadPath, showBrowser } from "./config.js"
import { parseCookieObject } from "./utils.js"
import { firefox as browserCore } from "playwright"
import { exec } from "child_process"

/* 使用方法
 * node uploadsubs.js <batchSize> <userId>
 * batchSize 扫描最近的视频数量 默认5条
 * userId    用户ID （可选）默认当前用户
 */
const batchSize = process.argv[2] || 5
const userId = process.argv[3] || bilibiliCookies.DedeUserID
main(batchSize, userId)

// 目标语言
const langSelectorMap = {
  en: {
    name: "English",
    langCode: "en",
    langSelector: "div.el-select-dropdown.el-popper > div > ul > li:nth-child(6)",
  },
  ja: {
    name: "日本語",
    langCode: "ja",
    langSelector: "div.el-select-dropdown.el-popper > div > ul > li:nth-child(7)",
  },
  zh: {
    name: "简体中文",
    langCode: "zh-Hans",
    langSelector: "div.el-select-dropdown.el-popper > div > ul > li:nth-child(1)",
  }
}

async function downloadSubtitle(youtubeUrl, bvid){
  return new Promise((resolve, reject) => {
    const command = `youtube-dl "${youtubeUrl}" --all-subs -o ${downloadPath}${bvid} --skip-download`
    console.log(command);
    exec(command,
    (error, stdout, stderr) => {
      if (error) {
          console.log(`error: ${error.message}`);
          return reject(error);
      }
      if (stderr) {
          console.log(`stderr: ${stderr}`);
          return reject(error);
      }
      console.log(`stdout: ${stdout}`);
      resolve()
  })
  })
}

async function getApi(url) {
  let data = new Uint8Array(0)
  function jsonEscape(str) {
    return str
      .replace(/\n/g, "\\\\n")
      .replace(/\r/g, "\\\\r")
      .replace(/\t/g, "\\\\t")
  }
  return new Promise((resolve) => {
    https.get(url, (res) => {
      res.on("data", (chunk) => {
        data += chunk
      })
      res.on("end", (_) => {
        resolve(JSON.parse(jsonEscape(data.toString())))
      })
    })
  })
}

function blockImageAndTracker(page) {
  return page.route("**/*", (route) => {
    if (route.request().resourceType() === "image") return route.abort()
    let isTracker = /(log-reporter.js)|(data.bilibili.com\/log)/.test(
      route.request().url()
    )
    if (isTracker) return route.abort()
    return route.continue()
  })
}

async function main(batchSize, userId) {
  const spaceData = await getApi(
    `https://api.bilibili.com/x/space/arc/search?mid=${userId}&pn=1&ps=${batchSize}`
  )
  var vlist = spaceData.data.list.vlist
  vlist = vlist.filter(
    (it) => it.subtitle === "" && it.description.includes("youtu")
  )
  if (vlist.length === 0) {
    console.log("No youtube video detected.")
  }
  console.log(`Scanning ${vlist.length} video`);

  for await (let video of vlist) {
    const youtubeUrl = video.description.split("\n")[0]
    await downloadSubtitle(youtubeUrl, video.bvid)
  }
  //Rename vtt to srt
  fs.readdirSync(downloadPath).forEach(it => {
    if(it.indexOf(".vtt") === -1) return
    fs.renameSync(`${downloadPath}${it}`, `${downloadPath}${it.replace("vtt","srt")}`)
  })

  const browser = await browserCore.launch({
    headless: !showBrowser,
  })
  const context = await browser.newContext({
    userAgent:
      "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.182 Safari/537.36 Edg/88.0.705.74",
  })

  context.addCookies(parseCookieObject(bilibiliCookies))

  const page = await context.newPage()
  await blockImageAndTracker(page)

  
  //上传字幕阶段
  for await (let video of vlist) {
    const subtitleMatchedFile = fs.readdirSync(downloadPath).filter(it => it.indexOf(video.bvid) !== -1)
    if (subtitleMatchedFile.length === 0) {
      console.log(`[${video.bvid}] No subtitle for ${video.title}`);
      continue
    }
    await page.goto(`https://www.bilibili.com/video/${video.bvid}`)
    const [bvid, cid] = await page.evaluate(() => [window.bvid, window.cid])
    for (const lang in langSelectorMap){
      const {name: langName,langCode, langSelector} = langSelectorMap[lang]
      const subtitlePath = `${downloadPath}${video.bvid}.${langCode}.srt`
      if (!fs.existsSync(subtitlePath)) {
        console.log(`[${bvid}] ${langName}subtitle file not found（${subtitlePath}）>>> SKIP`)
        continue
      }
      
      await page.goto(
        `https://account.bilibili.com/subtitle/edit/#/editor?bvid=${bvid}&cid=${cid}`,
        {
          waitUntil: "load",
        }
      )
      await Promise.all([
        page.reload({
          timeout: 20_000,
          waitUntil: "networkidle",
        }),
        page.waitForResponse(/subtitle_lan.json/),
      ])

      const langElement = await page.$(langSelector)
      const langElementText = await langElement.textContent()
      if(langElementText.includes("已发布")){
        console.log(`[${bvid}] ${langName} subtitle already exist >>> SKIP`);
        continue
      }

      await langElement.click()
      const [fileChooser] = await Promise.all([
        page.waitForEvent("filechooser"),
        page.click('text="上传字幕"'),
      ])
      await fileChooser.setFiles(subtitlePath)
      await page.waitForTimeout(3_000)
      await page.click('text="提交"')
      await page.waitForTimeout(5_000)
      console.log(`[${bvid}] ${langName} subtitle submitted`)
    }
  }

  await page.close()
  await context.close()
  await browser.close()
}